### ISSUE TITLE
RDFizer: NLP usage

### ISSUE THREAD

**Issue by agazzarini** [ALIADA/aliada-tool#76]
As part of the final release we have to attach, somewhere in the conversion pipeline, a dedicated component for doing NLP analysis on some terms extracted from records. The NER results will be stored as RDF triples that will enrich the owning records.

**Comment by agazzarini** [comment]
I'm trying the Stanford NER library [1] which seems interesting. The only annoying thing is that it requries Java 8 so, after trying, if we decide to go for it, we could:

a) put this library on a separate tomcat which run with Java 8 and provide the NER service through HTTP
b) upgrade our JVM and use the library directly within the RDFizer

**Comment by agazzarini** [comment]
@idoiamurua @scanbit  what do you think about the previous comment? I completed a first trial of that NER and, differently from other similar software (e.g. stanbol, opennlp), it works simply and without a big overhead. The product comes with several knowledge base dbs that are able to detect nouns and entities also in our records (I tried something in spanish and in hungarian) 

However, as I said, it requires Java 8. Do you have any specific trouble in your modules if we upgrade our JVM?

**Comment by idoiamurua** [comment]
Will we have any problems with Tomcat7 and Java 8?

**Comment by agazzarini** [comment]
No I don't think

**Comment by idoiamurua** [comment]
@eturienzo What about Struts library?

**Comment by agazzarini** [comment]
Theoretically there shouldn't be any problem at all. However, it would 
be great if we could try our modules with Java 8

In addition, the EOL of Java7 is scheduled for next April so it will 
have a short life.

Another alternative should be an additional tomcat providing those NER 
services through HTTP.

**Comment by idoiamurua** [comment]
Shall we change the Java version in the Aliada POM to 8?
To do so, Is it enough to change the following lines in the POM?

		<maven.compiler.source>1.7</maven.compiler.source>
		<maven.compiler.target>1.7</maven.compiler.target>
		<java.version>1.7</java.version>

**Comment by agazzarini** [comment]
Yes. That should be enough (assuming you have Java 8 installed)

**Comment by idoiamurua** [comment]
@scanbit @eturienzo @xmolero Could you make a try of the UI (struts) with Java 8?

**Comment by eturienzo** [comment]
I am going to check it with th UI. When I finish, I will tell you

**Comment by idoiamurua** [comment]
OK. I will try my code too.

**Comment by eturienzo** [comment]
We have no problem with the change to JAVA8.

**Comment by agazzarini** [comment]
Ok, @idoiamurua, if I remember the interlink uses java standalone 
processes...if you have problem with Java8 you could always run them 
with java 7

---------------------------

Regardless the path we will follow, let's assume that on top of 1 record 
I detect 2 names and 1 place: where do we put this information? Do we 
have to use the EFRBROO ontology? Because I think this information 
should be kept apart, they aren't controlled and maybe they could 
contains mistakes

**Comment by idoiamurua** [comment]
Yes. I do not think there is no problem with the Links Discovery Client application module.
I have just built all ALIADA modules with Java 8 in my PC and seems to give no problems.
I will change the Java version of the ALIADA POM in Github.

**Comment by agazzarini** [comment]
@cgareta @tpossemato could you please reply to this topic with a list of 
the  most representative tags we should consider as NER input?

I need this lists separated by input format, so one for MARC, one for 
LIDO and one for DC

thx

**Comment by agazzarini** [comment]
@idoiamurua I'd like to know your opinion about where to store the named entities extracted with NER...do you know if any of the imported ontologies do have a suitable predicate for that?

**Comment by agazzarini** [comment]
The knowledge model I'm using is able to detect 

- Person
- Organization
- Places
- Dates
- Time
- and other things like Money which we aren't interesting

Ideally, we should transform those entities in triples maintaining their type but I'm not sure how to proceed. The ontology doesn't seem to provide suitable predicates for such scenario

**Comment by idoiamurua** [comment]
I have sent an e-mail to Marta, she is not in the office today, about this issue. I proposed to her the following options:

 * CRM has the data property P3_has_note.

 * SKOS has the “Annotation Property” skos:scopeNote defined as :


       <owl:AnnotationProperty rdf:about="&skos;scopeNote">
            <rdfs:label xml:lang="en">scope note</rdfs:label>
            <definition xml:lang="en">A note that helps to clarify the meaning and/or the use of a concept.</definition>
             <rdfs:isDefinedBy rdf:resource="http://www.w3.org/2004/02/skos/core"/>
             <rdfs:subPropertyOf rdf:resource="&skos;note"/>
         </owl:AnnotationProperty>

I hope tomorrow I have an answer.

**Comment by agazzarini** [comment]
Great! Many thanks. How can we distinguish between a Person and Place?

**Comment by idoiamurua** [comment]
I have just been talking with Marta, because she confectionated Aliada ontology and we have been discussing about how can we add these new subjects.
She will look it more deeply, but now she is busy with other things. She hopes to answer by next Monday.
We have seen the next two other possible properties existing in Aliada ontology:

  * foaf:topic
         * domain foaf:Document . Problem: foaf_Document not aligned yet with ECRM/FRBRoo ontology and our used classes frbr:F3_Manifestation_Product_Type or ecrm:E19_Physical_Object.
          * range Thing
 * ecrm:P67_refers_to
          * domain ecrm:E89_Propositional_Object . Problem: our instances are of type frbr:F3_Manifestation_Product_Type or ecrm:E19_Physical_Object, not subclasses of ecrm:E89_Propositional_Object.
          * range ecrm:E1_CRM_Entity

**Comment by agazzarini** [comment]
Absolutely no rush at all. I have to prepare things in the codebase so 
the triple production is the last thing.ù

Many thanks for your help

**Comment by agazzarini** [comment]
About the ecrm:P67_refers_to 
I think it's not a problem is our entity don't belong to that doamin: it will when we say

:xyz ecrm:P67_refers_to E1_CRM_Entity

in this way, regardless what is xyz, it will became also an E89_Propositional_Object

What about the object? I mean imagine that I detected the place "Florence" ...how does the E1_CRM_Entity hold this literal value?  How can we typize that entity so we can distinguish that from another (person) "Alessandro Manzoni"?

**Comment by idoiamurua** [comment]
I have just made a trial creating the following two objects:


       <owl:NamedIndividual rdf:about="&efrbroo;NLPTrial_Object">
           <rdf:type rdf:resource="&ecrm;E19_Physical_Object"/>
           <rdf:type rdf:resource="&ecrm;E89_Propositional_Object"/>
       </owl:NamedIndividual>
      <owl:NamedIndividual rdf:about="&efrbroo;NLPTrial_Manif">
           <rdf:type rdf:resource="&ecrm;E89_Propositional_Object"/>
           <rdf:type rdf:resource="&efrbroo;F3_Manifestation_Product_Type"/>
       </owl:NamedIndividual>


and then I have passed the OWLMicroREasoner, and it has given me the following errors:


       Conflicts
            - Error ("conflict"): "Individual a member of disjoint classes"
       Culprit = http://erlangen-crm.org/efrbroo/NLPTrial_Object
       Implicated node: http://erlangen-crm.org/current/E18_Physical_Thing
       Implicated node: http://erlangen-crm.org/current/E28_Conceptual_Object

The two classes are disjoint, so it would not be possible. 
There would no problem with the F3_Manifestation_Product_Type.

**Comment by agazzarini** [comment]
This is another interesting question: what is the target entity? Work? 
Expression? Manifestation?

I think @cgareta or @tpossemato could help us

**Comment by idoiamurua** [comment]
To distinguish between a Person, an Organization or a Place, we should create instances of type ecrm:E21_Person, ecrm:E39_Actor or a E53_Place respectivelly. Then add the corresponding object properties: ecrm:P131_is_identified_by, ecrm:P131_is_identified_by and ecrm:P87_is_identified_by, respectivelly, where the literals the NLP processor has found would be inserted.

**Comment by cgareta** [comment]
Good question: work is the main entity, but catalogues are plenty of manifestations without tags at work level… Manifestation+item is the most common situation

 

De: Andrea Gazzarini [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 10:11
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

This is another interesting question: what is the target entity? Work? 
Expression? Manifestation?

I think @cgareta or @tpossemato could help us

**Comment by agazzarini** [comment]
Hi Cristina,
many thanks for your prompt response.

I think it would be great to provide a simple map between tag/subfield and
the corresponding entities. For instance,

tag 500$a = Manifestation
>

That means: the outcoming entities (from a NER service execution that
received as input the text content of 500$a) must be assigned to
Manifestation level

Best
Andrea

2015-03-02 12:40 GMT+01:00 cgareta <notifications@github.com>:

**Comment by tpossemato** [comment]
Hi Andrea, attached a document with a mapping of most used Marc21 tags for Work, Expression and Person/Family/Corporate body: all tags not included in the sheets, if coming from Marc21 bibliographic, has to be considered useful to identify manifestations.

So, generally speaking:

-              for Work, Expression and Person/Family/Corporate body: you need to consider at least authority data plus bibliographic data reported in the sheets;

-              - for manifestations: you need to consider above all bibliographic tags, with exception of the tags reported in the sheets for other entities.

Let me know if it’s useful to understand or if you prefer a new mapping schema.

Bye

Tiziana

 

 

Da: cgareta [mailto:notifications@github.com] 
Inviato: lunedì 2 marzo 2015 12:41
A: ALIADA/aliada-tool
Cc: tpossemato
Oggetto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Good question: work is the main entity, but catalogues are plenty of manifestations without tags at work level… Manifestation+item is the most common situation



De: Andrea Gazzarini [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 10:11
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)



This is another interesting question: what is the target entity? Work? 
Expression? Manifestation?

I think @cgareta or @tpossemato could help us

**Comment by agazzarini** [comment]
Tiziana, GitHub doesn't allow attachments in posts. Please send me the file
and I will submit to our repository

2015-03-02 12:51 GMT+01:00 tpossemato <notifications@github.com>:

**Comment by cgareta** [comment]
That’s is in the mapping I sent. Do you need something more simple? A 2-column table?

 

Best

Cristina

 

De: Andrea Gazzarini [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 12:45
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Hi Cristina,
many thanks for your prompt response.

I think it would be great to provide a simple map between tag/subfield and
the corresponding entities. For instance,

tag 500$a = Manifestation
>

That means: the outcoming entities (from a NER service execution that
received as input the text content of 500$a) must be assigned to
Manifestation level

Best
Andrea

2015-03-02 12:40 GMT+01:00 cgareta < <mailto:notifications@github.com> notifications@github.com>:

**Comment by cgareta** [comment]
Please, consider also the entities related to subjects:

 

-          Concept

-          Place

-          Event

-          Object

 

These entities are related to the work

 

Cristina

 

De: tpossemato [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 12:52
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Hi Andrea, attached a document with a mapping of most used Marc21 tags for Work, Expression and Person/Family/Corporate body: all tags not included in the sheets, if coming from Marc21 bibliographic, has to be considered useful to identify manifestations.

So, generally speaking:

- for Work, Expression and Person/Family/Corporate body: you need to consider at least authority data plus bibliographic data reported in the sheets;

- - for manifestations: you need to consider above all bibliographic tags, with exception of the tags reported in the sheets for other entities.

Let me know if it’s useful to understand or if you prefer a new mapping schema.

Bye

Tiziana





Da: cgareta [ <mailto:notifications@github.com> mailto:notifications@github.com] 
Inviato: lunedì 2 marzo 2015 12:41
A: ALIADA/aliada-tool
Cc: tpossemato
Oggetto: Re: [aliada-tool] RDFizer: NLP usage (#76)



Good question: work is the main entity, but catalogues are plenty of manifestations without tags at work level… Manifestation+item is the most common situation



De: Andrea Gazzarini [ <mailto:notifications@github.com> mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 10:11
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)



This is another interesting question: what is the target entity? Work? 
Expression? Manifestation?

I think @cgareta or @tpossemato could help us

**Comment by agazzarini** [comment]
Ok @cgareta, yes I see.

Now, before seeing the attachment of Tiziana, in your file there are listed
a lot of tags, but as you know, we don't have to process all tags with NLP;
for that kind of recognition we need only those tag that do have a
reasonable amount of text, like notes.

Could you please list them?

Best,
Andrea

2015-03-02 13:30 GMT+01:00 cgareta <notifications@github.com>:

**Comment by cgareta** [comment]
Are you going to process notes? Main notes are:

 


NOTE TYPE

MARC TAG

SUBFIELD

FRBR ENTITY

RELATIONSHIP


500 - General Note (R) Full | Concise

				

501 - With Note (R) Full | Concise

				

502 - Dissertation Note (R) Full | Concise

				

504 - Bibliography, etc. Note (R) Full | Concise

504

a

Work

Form of work


505 - Formatted Contents Note (R) Full | Concise

505

a

Expression

Summarization of content


506 - Restrictions on Access Note (R) Full | Concise

				

507 - Scale Note for Graphic Material (NR) Full | Concise

				

508 - Creation/Production Credits Note (R) Full | Concise

				

510 - Citation/References Note (R) Full | Concise

				

511 - Participant or Performer Note (R) Full | Concise

511

a

Manifestation

Statement of responsibility


513 - Type of Report and Period Covered Note (R) Full | Concise

				

514 - Data Quality Note (NR) Full | Concise

				

515 - Numbering Peculiarities Note (R) Full | Concise

				

516 - Type of Computer File or Data Note (R) Full | Concise

				

518 - Date/Time and Place of an Event Note (R) Full | Concise

				

520 - Summary, etc. (R) Full | Concise

520

a

Expression

Summarization of content


521 - Target Audience Note (R) Full | Concise

521

a

Work

Intended audience


522 - Geographic Coverage Note (R) Full | Concise

522

a

Place

Term for place


524 - Preferred Citation of Described Materials Note (R) Full | Concise

				

525 - Supplement Note (R) Full | Concise

				

526 - Study Program Information Note (R) Full | Concise

				

530 - Additional Physical Form available Note (R) Full | Concise

530

a

Manifestation

Alternate


533 - Reproduction Note (R) Full | Concise

				

534 - Original Version Note (R) Full | Concise

				

535 - Location of Originals/Duplicates Note (R) Full | Concise

				

536 - Funding Information Note (R) Full | Concise

				

538 - System Details Note (R) Full | Concise

				

540 - Terms Governing Use and Reproduction Note (R) Full | Concise

540

a

Manifestation

Access restrictions


541 - Immediate Source of Acquisition Note (R) Full | Concise

541

a

Item

Provenance


542 - Information Relating to Copyright Status (R) Full | Concise

				

544 - Location of Other Archival Materials Note (R) Full | Concise

				

545 - Biographical or Historical Data (R) Full | Concise

545

a

Person/Corp

Biography/history*


546 - Language Note (R) Full | Concise

546

a

Expression

Language of expression


547 - Former Title Complexity Note (R) Full | Concise

				

550 - Issuing Body Note (R) Full | Concise

				

552 - Entity and Attribute Information Note (R) Full | Concise

				

555 - Cumulative Index/Finding Aids Note (R) Full | Concise

				

556 - Information About Documentation Note (R) Full | Concise

				

561 - Ownership and Custodial History (R) Full | Concise

				

562 - Copy and Version Identification Note (R) Full | Concise

				

563 - Binding Information (R) Full | Concise

				

565 - Case File Characteristics Note (R) Full | Concise

				

567 - Methodology Note (R) Full | Concise

				

580 - Linking Entry Complexity Note (R) Full | Concise

				

581 - Publications About Described Materials Note (R) Full | Concise

				

583 - Action Note (R) Full | Concise

				

584 - Accumulation and Frequency of Use Note (R) Full | Concise

				

585 - Exhibitions Note (R) Full | Concise

				

586 - Awards Note (R) Full | Concise

				

588 - Source of Description Note (R) Full | Concise

				

59X - Local Notes Full | Concise

				

 

 

De: Andrea Gazzarini [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 13:37
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Ok @cgareta, yes I see.

Now, before seeing the attachment of Tiziana, in your file there are listed
a lot of tags, but as you know, we don't have to process all tags with NLP;
for that kind of recognition we need only those tag that do have a
reasonable amount of text, like notes.

Could you please list them?

Best,
Andrea

2015-03-02 13:30 GMT+01:00 cgareta < <mailto:notifications@github.com> notifications@github.com>:

**Comment by agazzarini** [comment]
Sorry we need another dimension: the source format, because tags are valid
for marc but we need that also for lido and dc, too

**Comment by cgareta** [comment]
Ok, I will try to finish the revisión of the mapping from dc, but I don’t know if LIDO includes notes like marc notes

 

 

De: Andrea Gazzarini [mailto:notifications@github.com] 
Enviado el: lunes, 02 de marzo de 2015 15:04
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Sorry we need another dimension: the source format, because tags are valid
for marc but we need that also for lido and dc, too

**Comment by tpossemato** [comment]
Andrea, see if the file I have sent you per e-mail  is what can help you: if yes, I'll continue tomorrow to complete the report.

**Comment by idoiamurua** [comment]
Here is the mapping to use to save the named entities extracted with NLP:

 * OPTION 1 – Work & Expressions

The extracted named entities can be instances of skos:Concept (in ALIADA ontology skos:Concept is subclass of E55_Type, and so it is a E1_CRM_Entity) . The Work/Expression would be an instance of E89_PropositionalObject, and they would be related as follows:

E89_PropositionalObject         P129_is_about            E1_CRM_Entity
(work, expression)                                                      (skos:concept)
  

* OPTION 2 – Manifestations, Items, PhysicalObjects of LIDO

The extracted named entities can be instances of skos:Concept (in ALIADA ontology skos:Concept is subclass of E55_Type) . So, they would be related as follows:

E1_CRM_Entity                                 P137 exemplifies    skos:Concept. 
(manifestation, ítem, LIDO physicalObject)                                          (skos:concept)

To differentiate among Persons, Places, etc. Marta says that we should not do it too complicated as I commented to you before. That is, we do not need to create instances of Person, Place, etc, but all the extracted named entities will be instances of skos:concept, and we will group them using the skos:Collection class.
That is, we will create the following collections:

       PersonsCollection rdf:type skos:Collection
       PlacesCollection rdf:type skos:Collection
       ????.....

When a named entity is extracted (e.g. a Person name like Darwin), the following triples are generated:

      Darwin rdf:type skos:Concept
      PersonsCollection skos:member Darwin
      Work1 rdf:type ecrm:E89_PropositionalObject 
      Work1 ecrm:P129_is_about Darwin

In the end, I do not know if you (Andrea, Cristina and Tizziana) agreed the level where to add the extracted named entities (at Work, Expression or Manifestation level), but the 2 options explained here let you do it at any of the levels you decide.

**Comment by cgareta** [comment]
Hi all, 

 

If we use only that level of entities we will lose a lot of attributes related to the manifestation (and finally to the bibliographic record because stored bibliographic records are usually manifestations), and the relationship to the item entity. I think that it’s not a good idea… but I have to read the explanation more slowly to be aware of the problem

 

@tiziana, what do you think about it?

 

Cristina

 

De: Idoia [mailto:notifications@github.com] 
Enviado el: miércoles, 04 de marzo de 2015 11:59
Para: ALIADA/aliada-tool
CC: cgareta
Asunto: Re: [aliada-tool] RDFizer: NLP usage (#76)

 

Here is the mapping to use to save the named entities extracted with NLP:

*	OPTION 1 – Work & Expressions

The extracted named entities can be instances of skos:Concept (in ALIADA ontology skos:Concept is subclass of E55_Type, and so it is a E1_CRM_Entity) . The Work/Expression would be an instance of E89_PropositionalObject, and they would be related as follows:

E89_PropositionalObject P129_is_about E1_CRM_Entity
(work, expression) (skos:concept)

*	OPTION 2 – Manifestations, Items, PhysicalObjects of LIDO

The extracted named entities can be instances of skos:Concept (in ALIADA ontology skos:Concept is subclass of E55_Type) . So, they would be related as follows:

E1_CRM_Entity P137 exemplifies skos:Concept. 
(manifestation, ítem, LIDO physicalObject) (skos:concept)

To differentiate among Persons, Places, etc. Marta says that we should not do it too complicated as I commented to you before. That is, we do not need to create instances of Person, Place, etc, but all the extracted named entities will be instances of skos:concept, and we will group them using the skos:Collection class.
That is, we will create the following collections:

   PersonsCollection rdf:type skos:Collection
   PlacesCollection rdf:type skos:Collection
   ????.....

When a named entity is extracted (e.g. a Person name like Darwin), the following triples are generated:

  Darwin rdf:type skos:Concept
  PersonsCollection skos:member Darwin
  Work1 rdf:type ecrm:E89_PropositionalObject 
  Work1 ecrm:P129_is_about Darwin

In the end, I do not know if you (Andrea, Cristina and Tizziana) agreed the level where to add the extracted named entities (at Work, Expression or Manifestation level), but the 2 options explained here let you do it at any of the levels you decide.

**Comment by idoiamurua** [comment]
@cgareta We are not going to remove anything. We are going to add new properties to the work, manifestation, item or whatever you decide to associate to the extracted named entity.

**Comment by agazzarini** [comment]
Hi @idoiamurua @marta, 
many thanks, sounds really good!

**Comment by idoiamurua** [comment]
Regarding LIDO, the NLP should be applied to the following data propery:

ecrm:E19_Physical_Object	ecrm:P3_has_note	  xsd:string

These triples will be generated from the following tags in the LIDO records:


      <lido:lidoWrap xmlns:lido="http://www.lido-schema.org">
          <lido:lido>
               <lido:descriptiveMetadata xml:lang="en">
                     <lido:objectIdentificationWrap>
                         <lido:objectDescriptionWrap>
                               <lido:objectDescriptionSet>
                                     <lido:descriptiveNoteValue xml:lang="en" lido:type ="physical-description"> Sculpture of Mozart …..</lido:descriptiveNoteValue >

**Comment by agazzarini** [comment]
Hi @idoiamurua sorry I lost your email about the question: few days ago you asked where we are going to put the literal for the extracted entities. Now that I'm implementing that stuff, I see that the things is wrapped as a concept. As consequence of that we should use skos:prefLabel instead of p3_has_note

Is that correct?

**Comment by idoiamurua** [comment]
Yes. I was about to answer to you that it was better to use skos:prefLabel. The difference with ecrm:P3_has_note is that skos:prefLabel is a annotation property and ecrm:P3_has_note is a data property.
The e-mail you were looking for is in the following issue: https://github.com/ALIADA/aliada-tool/issues/37

**Comment by agazzarini** [comment]
@idoiamurua I'm lost with the OPTION 2 above...could you please write down an example where the extracted entity is attached to a LIDO object?

**Comment by idoiamurua** [comment]
```
Darwin rdf:type skos:Concept 
PersonsCollection skos:member Darwin 
E19_Physical_Object1 ecrm:P137_exemplifies Darwin
Darwin skos:prefLabel "Darwing" . 

```
That is, you do not need to create any other rdf:type triple for the LIDO object, just add the ecrm:P137_exemplifies object property (the same applies to the manifestations).

**Comment by agazzarini** [comment]
+1 Many thanks!

**Comment by agazzarini** [comment]
As part of the last two pushes, the NLP branch (along with its feature) has been merged into master. 
I implemented that both in LIDO and in MARC. For this latter, I tried to detect entities only for few tags because that processing is very heavy, from a computational point of view. As consequence of that, I suggest: we should try to solve the performance issue raised by Adam and then increment the NER usage step by step, tag by tag.
